src/codegen/mod.rs

```diff
@@ -1160,35 +1177,57 @@ impl Module {
             .map(Clone::clone)
             .collect();
 
+        tests.into_iter().map(|name| TestCase {
+            name: name.replace("test#", ""),
+            func: self
+                .get_function::<Ctx, fn() -> Verdict<(), ()>>(
+                    name.strip_prefix("pkg.").unwrap(),
+                )
+                .unwrap(),
+        })
+    }
+
+    pub fn run_tests<Ctx: 'static>(&mut self, ctx: Ctx) -> Result<(), ()> {
+        self.run_tests_with_writer(ctx, &mut std::io::stdout())
+    }
+
+    pub fn run_tests_with_writer<Ctx: 'static>(
+        &mut self,
+        mut ctx: Ctx,
+        mut w: &mut impl std::io::Write,
+    ) -> Result<(), ()> {
+        let tests: Vec<_> = self.get_tests::<Ctx>().collect();
+
         let total = tests.len();
         let total_width = total.to_string().len();
         let mut successes = 0;
         let mut failures = 0;
 
         for (n, test) in tests.into_iter().enumerate() {
             let n = n + 1;
-            let test_display = test.replace("test#", "");
-            print!("Test {n:>total_width$} / {total}: {test_display}... ");
-            let test_fn = self
-                .get_function::<Ctx, fn() -> Verdict<(), ()>>(
-                    test.strip_prefix("pkg.").unwrap(),
-                )
-                .unwrap();
+            let test_display = test.name();
+            write!(
+                &mut w,
+                "Test {n:>total_width$} / {total}: {test_display}... "
+            )
+            .map_err(|_| ())?;
 
-            match test_fn.call(&mut ctx) {
-                Verdict::Accept(()) => {
+            match test.run(&mut ctx) {
+                Ok(()) => {
                     successes += 1;
-                    println!("\x1B[92mok\x1B[m");
+                    writeln!(&mut w, "\x1B[92mok\x1B[m").map_err(|_| ())?;
```

Comment:
Maybe make the return type `io::Result<Result<(), ()>>` and then just `?` these

Reply:
Another option would be to drop this part, keep using stdout, and if anyone wants to write to some other place: there's `.get_tests()`. The wrapping around it is - imo - small enough to be reimplemented.

I mean, there are ANSI colorcodes embedded here, which don't make much sense, unless the output is std{out,err}.

What do you think?

Reply:
If the need for a `run_tests_with_writer()` arises, it can always be added later, in a way that satisfies *those* needs better. I mean, it's not a particularly flexible function in the shape I implemented it... it was Just Enough™ to satisfy a need. As it happens, `.get_tests()` satisfies that need much better.

I'll remove the function then, and will change the title & description of the PR too, once I'm done with the rest of the cleanups.